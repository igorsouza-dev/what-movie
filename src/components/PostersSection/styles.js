import styled from 'styled-components';

export const Title = styled.h2``;
export const Container = styled.div`
  width: 100%;
  display: flex;
  flex-direction: column;
  margin-top: 40px;
`;
