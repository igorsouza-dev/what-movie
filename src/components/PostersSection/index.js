import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';

import Posters from 'components/Posters';
import api from 'services/api';
import { Container, Title } from './styles';

function PostersSection({ title, query, showButtons }) {
  const [movies, setMovies] = useState(null);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState('');
  useEffect(() => {
    async function getMovies() {
      try {
        const response = await api.get(query);
        const { results } = response.data;
        setMovies(
          results.map((movie) => {
            const url = movie.poster_path
              && `${process.env.REACT_APP_TMDB_IMAGE_URL}/w185${movie.poster_path}`;
            return { ...movie, url };
          }),
        );
      } catch (e) {
        setError(e.message);
      }
      setLoading(false);
    }
    getMovies();
  }, [query]);
  return (
    <Container>
      <Title>{title}</Title>
      {loading && <div>Loading...</div>}
      {movies
        && (movies.length ? (
          <Posters movies={movies} showButtons={showButtons} />
        ) : (
          <div>No movies found</div>
        ))}
      {error && <div>{error}</div>}
    </Container>
  );
}

PostersSection.propTypes = {
  title: PropTypes.string.isRequired,
  query: PropTypes.string.isRequired,
  showButtons: PropTypes.bool,
};
PostersSection.defaultProps = {
  showButtons: false,
};

export default PostersSection;
