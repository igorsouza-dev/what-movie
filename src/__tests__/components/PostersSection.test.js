import React from 'react';
import { createMemoryHistory } from 'history';
import { Router } from 'react-router-dom';
import {
  cleanup, render, waitForElement, waitForElementToBeRemoved
} from '@testing-library/react';
import apiMock from 'services/api';
import PostersSection from 'components/PostersSection';

jest.mock('services/api');


describe('<PostersSection />', () => {
  afterEach(cleanup);
  const title = 'Test section';
  const query = 'http://test.url.com';

  it('should render', async () => {
    const history = createMemoryHistory();
    apiMock.get.mockResolvedValueOnce(
      {
        data: {
          results: [
            {
              id: 1234,
              original_title: 'Movie name test',
              poster_path: '123456.jpg',
              average_vote: 10
            },
          ],
        }
      }
    );
    const { container, getByText, getByTestId } = render(
      <Router history={history}>
        <PostersSection title={title} query={query} />
      </Router>
    );

    expect(getByText(title)).toBeInTheDocument();

    expect(apiMock.get).toHaveBeenCalledWith(query);

    expect(getByText('Loading...')).toBeInTheDocument();

    await waitForElement(() => getByTestId(/poster-img-container/i));

    expect(getByText('Movie name test')).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });
  it('should render a message of empty results', async () => {
    const history = createMemoryHistory();
    apiMock.get.mockResolvedValueOnce(
      {
        data: {
          results: [

          ],
        }
      }
    );
    const { container, getByText } = render(
      <Router history={history}>
        <PostersSection title={title} query={query} />
      </Router>
    );

    expect(getByText(title)).toBeInTheDocument();

    expect(apiMock.get).toHaveBeenCalledWith(query);

    expect(getByText('Loading...')).toBeInTheDocument();

    await waitForElementToBeRemoved(() => getByText('Loading...'));

    expect(getByText('No movies found')).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });
});
