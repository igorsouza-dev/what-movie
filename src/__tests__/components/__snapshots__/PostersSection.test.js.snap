// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PostersSection /> should render 1`] = `
<div
  class="sc-jzJRlG fxCPVP"
>
  <h2
    class="sc-fjdhpX kwQVFY"
  >
    Test section
  </h2>
  <div
    class="sc-jTzLTM gjILoC"
    type="flex"
  >
    <div
      class="sc-ifAKCX cRQFzp"
    >
      <div
        class="sc-bZQynM mjAAS"
      >
        /10
      </div>
      <div
        class="sc-VigVT iOQNnG"
        data-testid="poster-img-container"
      >
        <img
          alt="Movie name test"
          src="https://image.tmdb.org/t/p/w185123456.jpg"
        />
      </div>
      <div
        class="sc-gzVnrw dFkHji"
      >
        <div
          class="sc-EHOje IcjOp"
          data-testid="name-container"
        >
          <strong>
            Movie name test
          </strong>
        </div>
      </div>
    </div>
  </div>
  
</div>
`;

exports[`<PostersSection /> should render a message of empty results 1`] = `
<div
  class="sc-jzJRlG fxCPVP"
>
  <h2
    class="sc-fjdhpX kwQVFY"
  >
    Test section
  </h2>
  <div>
    No movies found
  </div>
  
</div>
`;
